File: apps/rhino-cli/cmd/env_init.go
package cmd

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"

	"github.com/spf13/cobra"
)

var envInitForce bool

var envInitCmd = &cobra.Command{
	Use:   "init",
	Short: "Create .env files from .env.example templates",
	Long: `Finds all .env.example files in infra/dev/ and copies them to .env
in the same directory. Existing .env files are not overwritten unless --force is used.

This is useful for fresh setups where no rhino-cli env backup exists.`,
	Args:          cobra.NoArgs,
	SilenceErrors: true,
	RunE:          runEnvInit,
}

func init() {
	envCmd.AddCommand(envInitCmd)
	envInitCmd.Flags().BoolVar(&envInitForce, "force", false, "Overwrite existing .env files")
}

// envInitWalkDir is overridable for testing.
var envInitWalkDir = filepath.WalkDir

// envInitReadFile is overridable for testing.
var envInitReadFile = os.ReadFile

// envInitWriteFile is overridable for testing.
var envInitWriteFile = os.WriteFile

// envInitStat is overridable for testing.
var envInitStat = os.Stat

func runEnvInit(cmd *cobra.Command, _ []string) error {
	repoRoot, err := findGitRoot()
	if err != nil {
		return fmt.Errorf("failed to find git repository root: %w", err)
	}

	infraDevDir := filepath.Join(repoRoot, "infra", "dev")

	var created, skipped int
	var errs []string

	walkErr := envInitWalkDir(infraDevDir, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		if d.Name() != ".env.example" {
			return nil
		}

		dir := filepath.Dir(path)
		envPath := filepath.Join(dir, ".env")
		relPath, _ := filepath.Rel(repoRoot, envPath)

		if !envInitForce {
			if _, statErr := envInitStat(envPath); statErr == nil {
				fmt.Fprintf(cmd.OutOrStdout(), "Skipped: %s (already exists, use --force to overwrite)\n", relPath)
				skipped++
				return nil
			}
		}

		data, readErr := envInitReadFile(path)
		if readErr != nil {
			errs = append(errs, fmt.Sprintf("failed to read %s: %v", path, readErr))
			return nil
		}

		if writeErr := envInitWriteFile(envPath, data, 0o644); writeErr != nil {
			errs = append(errs, fmt.Sprintf("failed to write %s: %v", envPath, writeErr))
			return nil
		}

		fmt.Fprintf(cmd.OutOrStdout(), "Created: %s (from %s)\n", relPath, filepath.Base(path))
		created++
		return nil
	})

	if walkErr != nil {
		return fmt.Errorf("failed to walk infra/dev/: %w", walkErr)
	}

	fmt.Fprintf(cmd.OutOrStdout(), "\nSummary: %d created, %d skipped\n", created, skipped)

	for _, e := range errs {
		fmt.Fprintf(cmd.ErrOrStderr(), "Error: %s\n", e)
	}

	return nil
}

File: apps/rhino-cli/src/commands/env_init.rs
//! `env init` — copies `.env.example` files to `.env` files under `infra/dev/` and `apps/`.
//!
//! Port of `apps/rhino-cli/cmd/env_init.go`.

use std::fs;
use std::path::{Path, PathBuf};

use anyhow::{Error, anyhow};
use clap::Args;
use walkdir::WalkDir;

use crate::domain::cliout::OutputFormat;
use crate::internal::git;

/// Directories under `repo_root` that are scanned for `.env.example` files.
const SCAN_ROOTS: &[&str] = &["infra/dev", "apps"];

/// Collect all `.env.example` files found under `SCAN_ROOTS` in `repo_root`.
pub fn collect_examples(repo_root: &Path) -> Vec<PathBuf> {
    let mut results = Vec::new();
    for root in SCAN_ROOTS {
        let scan_dir = repo_root.join(root);
        for entry in WalkDir::new(&scan_dir).into_iter().flatten() {
            if entry.file_type().is_dir() {
                continue;
            }
            if entry.file_name() == ".env.example" {
                results.push(entry.path().to_owned());
            }
        }
    }
    results
}

/// CLI arguments for `env init`.
#[derive(Args, Debug)]
pub struct EnvInitArgs {
    /// Overwrite existing .env files.
    #[arg(long = "force")]
    pub force: bool,
}

/// Run the `env init` command.
///
/// # Errors
///
/// Returns an error if the git root cannot be found or if a path operation
/// fails unexpectedly.
///
/// # Panics
///
/// Panics if a `walkdir` entry path has no filename component, which cannot
/// happen for entries produced by `WalkDir`.
pub fn run(args: &EnvInitArgs, _output: OutputFormat) -> std::result::Result<(), Error> {
    let repo_root =
        git::root::find_root().map_err(|e| anyhow!("failed to find git repository root: {e}"))?;
    let mut created = 0usize;
    let mut skipped = 0usize;
    let mut errs: Vec<String> = Vec::new();

    for path in collect_examples(&repo_root) {
        let env_path = path
            .parent()
            .ok_or_else(|| anyhow!("invalid path"))?
            .join(".env");
        let rel = env_path.strip_prefix(&repo_root).unwrap_or(&env_path);
        if !args.force && env_path.exists() {
            println!(
                "Skipped: {} (already exists, use --force to overwrite)",
                rel.display()
            );
            skipped += 1;
            continue;
        }
        let data = match fs::read(&path) {
            Ok(d) => d,
            Err(e) => {
                errs.push(format!("failed to read {}: {e}", path.display()));
                continue;
            }
        };
        if let Err(e) = fs::write(&env_path, data) {
            errs.push(format!("failed to write {}: {e}", env_path.display()));
            continue;
        }
        println!(
            "Created: {} (from {})",
            rel.display(),
            path.file_name()
                .expect("walkdir entry always has file_name")
                .to_string_lossy()
        );
        created += 1;
    }

    println!("\nSummary: {created} created, {skipped} skipped");
    for e in &errs {
        eprintln!("Error: {e}");
    }
    Ok(())
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::panic)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn make_fixture(tmp: &TempDir, rel: &str) {
        let p = tmp.path().join(rel);
        fs::create_dir_all(p.parent().unwrap()).unwrap();
        fs::write(&p, b"# fixture").unwrap();
    }

    #[test]
    fn args_default() {
        let _ = EnvInitArgs { force: false };
    }

    #[test]
    fn collect_examples_finds_apps_env_example() {
        let tmp = TempDir::new().unwrap();
        make_fixture(&tmp, "infra/dev/organiclever/.env.example");
        make_fixture(&tmp, "apps/ose-be/.env.example");

        let found: Vec<String> = collect_examples(tmp.path())
            .into_iter()
            .map(|p| {
                p.strip_prefix(tmp.path())
                    .unwrap()
                    .to_string_lossy()
                    .replace('\\', "/")
            })
            .collect();

        assert!(
            found.contains(&"apps/ose-be/.env.example".to_string()),
            "apps/ose-be/.env.example must be discovered; got: {found:?}"
        );
    }
}
